Issue 39 support othellogpt in SAELens

decandido opened the pull request:
# Description

Adding support for OthelloGPT. Adding logic to train on sequences with a start and end position offset similar to [PR 294](https://github.com/jbloomAus/SAELens/pull/294). Adding tests for the added logic and to benchmark the OthelloGPT SAETrainingRunner.

Fixes # Issue [39](https://github.com/jbloomAus/SAELens/issues/39)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

<!--
As you go through the checklist above, you can mark something as done by putting an x character in it

For example,
- [x] I have done this task
- [ ] I have not done this task
-->

### You have tested formatting, typing and unit tests (acceptance tests not currently in use)

- [x] I have run `make check-ci` to check format and linting. (you can run `make format` to format code if needed.)

jbloomAus commented:
Thanks @decandido and @zhenningdavidliu!

This has some overlap with https://github.com/jbloomAus/SAELens/pull/294. Specifically, I'd like to use the seqpos slicing instead of the start and end pos offsets. Other than that this looks good I think!

Requests:
- [] Please rebase or merge with https://github.com/jbloomAus/SAELens/pull/294 (which will also make sure the seqpos slice ends up in the config)
- [] Optional: Do you want to make a short tutorial on training / using the OthelloGPT SAE? 

Once that's done I'll accept :)

chanind commented on sae_lens/training/activations_store.py:
```diff
@@ -451,27 +455,42 @@ def get_activations(self, batch_tokens: torch.Tensor):
 
-        stacked_activations = torch.zeros((n_batches, n_context, 1, self.d_in))
+        # For some models, we might want to exclude some positions from the sequence to train on
+        training_context_slice = list(
```
This is inefficient to create a list of indices like this. It's better to create an actual `slice` object, like `training_context_slice = slice(self.start_pos_offset, n_context - self.end_pos_offset)`. Alternative, you can probably just do something like the following directly when selecting the specific indices.

```python
stacked_activations[:, :, 0] = layerwise_activations[self.hook_name][
      :, start_pos:end_pos, self.hook_head_index
]
```

decandido replied:
Hey David, thanks for your thorough review of our PR! As Joseph mentioned in his [comment](https://github.com/jbloomAus/SAELens/pull/317#issuecomment-2391999612) above, we've rebased on Callum's PR and added a few tests of our own. These problems are not present in the rebased code, but they are really helpful for @zhenningdavidliu and me! 🙏

chanind commented on sae_lens/training/activations_store.py:
```diff
@@ -491,10 +510,15 @@ def get_buffer(
+        training_context_slice = list(
+            range(self.start_pos_offset, context_size - self.end_pos_offset)
+        )
+        training_context_size = len(training_context_slice)
```
Here as well, this is inefficient to create a list from a range and then measure it. This would be better to just directly compute like `training_context_size = context_size - self.end_pos_offset - self.start_pos_offset`. This could also be moved to a shared function since it looks like it's shared with code in `get_activations()`

chanind commented on tests/unit/training/test_cache_activations_runner.py:
```diff
@@ -216,3 +216,89 @@ def W_E(self) -> torch.Tensor:
+
+    tokens_in_buffer = n_batches_in_buffer * store_batch_size * training_context_size
+    total_training_tokens = n_buffers * tokens_in_buffer
+    print(f"Total Training Tokens: {total_training_tokens}")
```
these print statements should be removed for tests

chanind commented on sae_lens/training/activations_store.py:
```diff
@@ -451,27 +455,42 @@ def get_activations(self, batch_tokens: torch.Tensor):
                 stacked_activations[:, :, 0] = layerwise_activations[
                     self.hook_name
-                ].view(n_batches, n_context, -1)
+                ].view(n_batches, training_context_size, -1)[
```
This looks like it's not correct - don't we want to first stack into `n_context` sequences and then lop off the start and end? It looks like as this is written it will turn this into `training_context_size` sequences and then further slice them, likely leading to the wrong size sequences. Is this covered by a test?

chanind reviewed (commented):
If I userstand this correctly, these offset parameters only makes sense in cases where the model is given sequences of a fixed length, and `context_size` is set to that length, e.g. OthelloGPT?

decandido commented:
> Requests:
> 
> * [] Please rebase or merge with [Support seqpos slicing #294](https://github.com/jbloomAus/SAELens/pull/294) (which will also make sure the seqpos slice ends up in the config)
> * [] Optional: Do you want to make a short tutorial on training / using the OthelloGPT SAE?

Thanks @jbloomAus for reviewing our PR! We've rebased on #294 and removed the start and end pos offsets. 

We also added a short script to train SAEs on othelloGPT (`scripts/training_a_sparse_autoencoder_othelloGPT.py`). Let us know if that's what you were looking for!

callummcdougall commented:
Hey, thanks for all your awesome work here! Could I possibly bump on this PR? Just cause it would also be super useful for the final bits of the ARENA material to run smoothly (-:

callummcdougall commented:
Thanks! (I'm actually also realizing that the context length issue wasn't a part of this PR, it was an earlier one, and that was the main thing that was making things a bit clunky, so less important thank I thought - but would still be great!)
